### ISSUE TITLE
SQLAlchemy Runtime Inspection API doesn't support `AsyncEngine`

### ISSUE THREAD

**Issue by Ahsoka** [sqlalchemy/sqlalchemy#6121]
The title pretty much describes this feature request.  I would greatly appreciate if SQLAlchemy would support the Runtime Inspection API with a `AsyncEngine` much in the same way that it does with normal `Engine` objects.  I know that the SQLAlchemy devs are very busy atm, so please don't feel any pressure to implement this soon.  Just wanted to let you guys know about this.

**Have a nice day!**

**Comment by zzzeek** [comment]
hey there.

it's not clear what inspect() would return for AsyncEngine.   I guess we'd have to build an async facade over Inspector - but this is difficult because dialects can provide subclasses of Inspector with additional methods, and those would not have an async version.

It's generally best if you inspect(async_engine.engine) and then use the sync API with the returned Inspector, and we can document this.    Maybe there can be an async facade for Inspector but I dont know how it would handle dialect-custom methods like Postgresql get_enums().

**Comment by zzzeek** [comment]
it would have to be a `__getattr__` scheme that returns an async facade over methods.   that would work but it wouldn't show up in mypy.

**Comment by zzzeek** [comment]
we can likely add this at some point and it can be sqlalchemy.ext.asyncio.reflection.   it would cover the Inspector API using proxies and for per-dialect inspectors would use a `__getattr__` scheme.

for now using run_sync() is the workaround.

**Comment by whg517** [comment]
Hello .

When i use `inspect` , and got a error `sqlalchemy.exc.NoInspectionAvailable: No inspection system is available for object of type <class 'sqlalchemy.ext.asyncio.engine.AsyncConnection'>`

code:

```python
@pytest.mark.asyncio
async def test_migrate(session):
    """
    Test migrate.
    :param session:
    :return:
    """
    async with session.bind.connect() as connection:
        inspector = inspect(connection)
        table_names = inspector.get_table_names()
        assert table_names
        assert 'audit' in table_names
```

In Async mode, how do I query all tables in the database? `SHOW TABLES` does not apply to sqlite3. Is there another good way to write it. `session.bind.dialect.get_table_names(connection)` is also not available

**Comment by Ahsoka** [comment]
> we can likely add this at some point and it can be sqlalchemy.ext.asyncio.reflection. it would cover the Inspector API using proxies and for per-dialect inspectors would use a `__getattr__` scheme.
> 
> for now using run_sync() is the workaround.

Hello @whg517.

As @zzzeek mentioned above the best solution would be to use the `run_sync()` method.  Consider the following example:
```py
@pytest.mark.asyncio
async def test_migrate(session):
    """
    Test migrate.
    :param session:
    :return:
    """
    def get_table_names(conn);
        inspector = inspect(conn)
        return inspector.get_table_names()

    async with session.bind.connect() as connection:
        table_names = connection.run_sync(get_table_names)
        assert table_names
        assert 'audit' in table_names
```

Learn more about the `run_sync` method here: https://docs.sqlalchemy.org/en/latest/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.run_sync

**Comment by CaselIT** [comment]
@zzzeek maybe this example could be documented in the asyncio documentation until we have an alternative solution?

**Comment by whg517** [comment]
I think this may be more concise, but it may not be as flexible as using `inspect`

```python
@pytest.mark.asyncio
async def test_migrate(session):
    """
    Test migrate.
    :param session:
    :return:
    """
    async with session.bind.connect() as connection:
        table_names = await connection.run_sync(session.bind.dialect.get_table_names)
        assert table_names
        assert 'audit' in table_names
```

**Comment by CaselIT** [comment]
I general the dialect method are intended to be used only by dialect author, and should not be used as general methods

**Comment by zzzeek** [comment]
> @zzzeek maybe this example could be documented in the asyncio documentation until we have an alternative solution?

yeah there should be a little section on "using the inspector" and there also should be some notes on how the event hooks work for Core and ORM.

**Comment by harshitsinghai77** [comment]
Update:

I got the workaround...

```python
engine = create_async_engine(url)

def load_table(conn):
        metadata = MetaData(bind=conn)
        metadata.reflect(only=["harshit_table"])
        harshit_table = Table("harshit_table", metadata, autoload_with=engine)
        return harshit_table

async def async_main(self):
        async with engine.connect() as conn:
            tables = await conn.run_sync(load_table)
            print("tables: ", tables)
````

Don't create metadata and use it like.
```python
engine = create_async_engine(url)
metadata = MetaData(bind=engine)

my_table = Table("harshit_table", metadata, autoload_with=engine)
```

This won't work.
